File: apps/api/src/routes/stream.ts
import { Hono } from "hono";
import { Env, Variables } from "../types";
import { query, queryOne, execute, apiResponse, successResponse, errorResponse, internalErrorResponse } from "../lib/db";

const app = new Hono<{ Bindings: Env; Variables: Variables }>();

const MAX_TOKENS_PER_HOUR = 10;
const userTokenCount: Record<string, number[]> = {};

function cleanupUserTokens(userId: string): number[] {
  const now = Date.now() / 1000;
  const tokens = userTokenCount[userId] || [];
  const recent = tokens.filter((t) => t > now - 3600);
  userTokenCount[userId] = recent;
  return recent;
}

function trialUnlockedCount(course: any): number {
  return Math.max(1, Math.ceil((course.syllabus || []).length * 0.1));
}

function trialActive(user: any): boolean {
  if (!user.trial_active) return false;
  const expires = user.trial_expires;
  if (!expires) return false;
  try {
    const expiresDate = new Date(expires);
    if (isNaN(expiresDate.getTime())) return false;
    return Date.now() < expiresDate.getTime();
  } catch {
    return false;
  }
}

async function hasAccess(
  env: Env,
  user: any,
  course: any,
  lessonIndex: number
): Promise<boolean> {
  if (user.role === "admin") return true;
  if (trialActive(user) && lessonIndex < trialUnlockedCount(course)) {
    return true;
  }
  const sub = await queryOne<any>(
    env,
    "SELECT * FROM subscriptions WHERE user_id = ? AND status = 'active'",
    [user._id]
  );
  if (!sub) return false;
  try {
    const endsAt = new Date(sub.ends_at);
    if (isNaN(endsAt.getTime())) return false;
    return Date.now() < endsAt.getTime();
  } catch {
    return false;
  }
}

app.get("/stream", async (c) => {
  try {
    const authHeader = c.req.header("Authorization");
    if (!authHeader?.startsWith("Bearer ")) {
      return c.json({ success: false, data: null, error: "Unauthorized", meta: null }, 401);
    }

    const token = authHeader.slice(7);
    const payload = c.get("user");
    if (!payload) {
      return c.json({ success: false, data: null, error: "Invalid token", meta: null }, 401);
    }

    const streamToken = c.req.query("token");
    if (!streamToken) {
      return c.json({ success: false, data: null, error: "Stream token required", meta: null }, 400);
    }

    const lessonId = await c.env.CACHE.get(`stream:${streamToken}`);
    if (!lessonId) {
      return c.json({ success: false, data: null, error: "Invalid or expired stream token", meta: null }, 401);
    }

    const video = await c.env.R2_BUCKET.get(`videos/${lessonId}.mp4`);
    if (!video) {
      return c.json({ success: false, data: null, error: "Video not found", meta: null }, 404);
    }

    return new Response(video.body, {
      headers: {
        "Content-Type": "video/mp4",
        "Cache-Control": "public, max-age=3600",
      },
    });
  } catch (error) {
    return internalErrorResponse(error);
  }
});

app.post("/lessons/:lesson_id/stream-token", async (c) => {
  try {
    const lessonId = c.req.param("lesson_id");
    const user = c.get("user");
    if (!user) {
      return c.json({ success: false, data: null, error: "Unauthorized", meta: null }, 401);
    }

    const courses = await query<any>(c.env, "SELECT * FROM courses");
    let course: any = null;
    let lesson: any = null;
    let lessonIndex = -1;

    for (const c of courses) {
      const syllabus = c.syllabus || [];
      for (let i = 0; i < syllabus.length; i++) {
        if (syllabus[i].id === lessonId) {
          course = c;
          lesson = syllabus[i];
          lessonIndex = i;
          break;
        }
      }
      if (course) break;
    }

    if (!course || !lesson) {
      return c.json({ success: false, data: null, error: "Lesson not found", meta: null }, 404);
    }

    const dbUser = await queryOne<any>(c.env, "SELECT * FROM users WHERE _id = ?", [user.sub]);
    if (!dbUser) {
      return c.json({ success: false, data: null, error: "User not found", meta: null }, 404);
    }

    if (!(await hasAccess(c.env, dbUser, course, lessonIndex))) {
      return c.json({ success: false, data: null, error: "Subscription or trial required", meta: null }, 403);
    }

    const userTokens = cleanupUserTokens(user.sub);
    if (userTokens.length >= MAX_TOKENS_PER_HOUR) {
      return c.json(
        { success: false, data: null, error: `Rate limit exceeded. Max ${MAX_TOKENS_PER_HOUR} tokens per hour.`, meta: null },
        429
      );
    }

    const r2Key = lesson.r2_key || lesson.drive_file_id;
    if (!r2Key) {
      return c.json({ success: false, data: null, error: "No video file associated with this lesson", meta: null }, 404);
    }

    const streamToken = `stream-${lessonId}-${Date.now()}-${Math.random().toString(36).substring(7)}`;
    await c.env.CACHE.put(`stream:${streamToken}`, lessonId, {
      expirationTtl: 3600,
    });

    userTokenCount[user.sub] = [...userTokens, Date.now() / 1000];

    return c.json(
      apiResponse(
        true,
        { stream_token: streamToken, lesson_id: lessonId },
        null,
        null
      ),
      200
    );
  } catch (error) {
    return internalErrorResponse(error);
  }
});

export default app;

File: apps/api/app/api/v1/stream.py
import math
import time
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException
from app.core.config import settings
from app.core.deps import get_current_user
from app.db.mongodb import get_db
from app.services.r2_storage import r2_storage

router = APIRouter()

# Per-user token rate limiting
_user_token_count: dict[str, list[float]] = {}
MAX_TOKENS_PER_HOUR = 10

DEMO_VIDEO_URL = "https://www.w3schools.com/html/mov_bbb.mp4"

logger = __import__("logging").getLogger(__name__)

def _cleanup_user_tokens(user_id: str):
    now = time.time()
    tokens = _user_token_count.get(user_id, [])
    tokens = [t for t in tokens if t > now - 3600]
    _user_token_count[user_id] = tokens
    return tokens


def _trial_unlocked_count(course: dict) -> int:
    return max(1, math.ceil(len(course.get("syllabus", [])) * 0.1))


def _trial_active(user: dict) -> bool:
    if not user.get("trial_active"):
        return False
    expires = user.get("trial_expires")
    if not expires:
        return False
    try:
        expires_dt = datetime.fromisoformat(expires)
        if expires_dt.tzinfo is None:
            expires_dt = expires_dt.replace(tzinfo=timezone.utc)
        return datetime.now(timezone.utc) < expires_dt
    except Exception:
        return False


async def _has_access(user: dict, course: dict, lesson_index: int, db) -> bool:
    if user.get("role") == "admin":
        return True
    if settings.bypass_subscription_check:
        return True
    if _trial_active(user) and lesson_index < _trial_unlocked_count(course):
        return True
    now_iso = datetime.now(timezone.utc).isoformat()
    sub = await db.subscriptions.find_one({
        "user_id": user["id"],
        "status": "active",
        "ends_at": {"$gt": now_iso},
    })
    if not sub:
        logger.info(
            "Stream access denied: user=%s reason=no_active_subscription lesson_index=%s course=%s",
            user.get("id"),
            lesson_index,
            course.get("_id"),
        )
        return False
    return True


@router.post("/lessons/{lesson_id}/stream-token")
async def create_stream_token(lesson_id: str, user: dict = Depends(get_current_user)):
    db = get_db()
    course = None
    lesson = None
    lesson_index = -1
    async for c in db.courses.find():
        for idx, l in enumerate(c.get("syllabus", [])):
            if l["id"] == lesson_id:
                course = c
                lesson = l
                lesson_index = idx
                break
        if course:
            break

    if not course or not lesson:
        raise HTTPException(status_code=404, detail="Lesson not found")

    if not await _has_access(user, course, lesson_index, db):
        raise HTTPException(status_code=403, detail="Subscription or trial required")

    user_tokens = _cleanup_user_tokens(user["id"])
    if len(user_tokens) >= MAX_TOKENS_PER_HOUR:
        raise HTTPException(
            status_code=429,
            detail=f"Rate limit exceeded. Max {MAX_TOKENS_PER_HOUR} tokens per hour.",
        )

    r2_key = lesson.get("r2_key")
    if not r2_key:
        drive_file_id = lesson.get("drive_file_id")
        if drive_file_id:
            # Legacy lesson with only Drive ID — trigger on-the-fly migration
            from app.services.watermark import migrate_drive_to_r2
            try:
                r2_key = await migrate_drive_to_r2(lesson_id, drive_file_id)
                syllabus = course.get("syllabus", [])
                for l in syllabus:
                    if l["id"] == lesson_id:
                        l["r2_key"] = r2_key
                        break
                await db.courses.update_one({"_id": course["_id"]}, {"$set": {"syllabus": syllabus}})
            except Exception:
                raise HTTPException(status_code=502, detail="Video migration from Drive failed")
        else:
            _user_token_count.setdefault(user["id"], []).append(time.time())
            return {
                "stream_url": DEMO_VIDEO_URL,
                "expires_in": 86400,
            }

    signed_url = await r2_storage.generate_signed_url(
        lesson_id, expires_in=settings.r2_signed_url_expiry_seconds
    )
    if not signed_url:
        raise HTTPException(status_code=502, detail="Failed to generate playback URL")

    _user_token_count.setdefault(user["id"], []).append(time.time())

    return {
        "stream_url": signed_url,
        "expires_in": settings.r2_signed_url_expiry_seconds,
    }
